### PR TITLE
UCP: Adding support for implicit non-blocking put/get

### PR DESCRIPTION
**Pull request by shamisp**
This is preliminary implementations for non-blocking put/get
operations.

### CODE REVIEW

**Comment by shamisp** [comment]
autobot build

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1197/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/618/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -1024,6 +1024,32 @@ ucs_status_ptr_t ucp_tag_msg_recv_nb(ucp_worker_h worker, void *buffer,
...
+ * @param [in]  ep           Remote endpoint handle.
+ * @param [in]  buffer       Pointer to the local source address.
+ * @param [in]  length       Length of the data (in bytes) stored under the
+ *                           source address.
+ * @param [in]  remote_addr  Pointer to the destination remote address
+ *                           to write to.
+ * @param [in]  rkey         Remote memory key associated with the
+ *                           remote address.
+ *
+ * @return Error code as defined by @ref ucs_status_t
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: Do we return UCS_OK if it was completed and UCS_INPROGRESS if started?

File: src/ucp/core/ucp_request.h [hunk 2]
```diff
@@ -68,6 +69,7 @@ typedef struct ucp_send_state {
  */
 typedef struct ucp_request {
     uint16_t                      flags;   /* Request flags */
+    uct_pending_req_t             uct;     /* Pending request */
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: we don't need this for 'recv' - no need to enlarge the request structure

File: src/ucp/core/ucp_request.h [hunk 3]
```diff
@@ -93,10 +95,18 @@ typedef struct ucp_request {
...
         } send;
 
         struct {
+            ucp_ep_h              ep;
+            const void            *buffer;     /* Put buffer */
+            uint64_t              remote_addr; /* remote address */
+            ucp_rkey_h            rkey;        /* Rkey */
+            size_t                length;      /* Put/Get length, in bytes */
+            ucp_frag_state_t      state;
+        } rma;
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: i'd suggest to add a union between 'rma' and 'tag' inside 'send'

File: src/ucp/rma/basic_rma.c [hunk 4]
```diff
@@ -75,6 +77,156 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
...
+
+    if (length == 0) {
+        /* Make sure that we mark is as a released */
+        ucp_request_complete(req, req->cb.send, UCS_OK);
+    }
+
+    return status;
+}
+
+#define ADD_PENDING_RMA(_req, _ep, _buf, _len, _raddr, _rkey, _cb) \
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: can this be inline function instead?

File: src/ucp/rma/basic_rma.c [hunk 5]
```diff
@@ -128,6 +280,101 @@ ucs_status_t ucp_get(ucp_ep_h ep, void *buffer, size_t length,
...
+            /* Make sure that we mark is as a released */
+            ucp_request_complete(req, req->cb.send, UCS_OK);
+            break;
+        } else if (ret < 0) {
+            break;
+        }
+    }
+    return ret;
+}
+
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: single spaceline here

File: src/ucp/rma/basic_rma.c [hunk 6]
```diff
@@ -128,6 +280,101 @@ ucs_status_t ucp_get(ucp_ep_h ep, void *buffer, size_t length,
...
+static ucs_status_t ucp_progress_get_nbi(uct_pending_req_t *self)
+{
+    int ret;
+
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, uct);
+
+    ucp_ep_t   *ep         = req->rma.ep;
+    const void       *buffer     = req->rma.buffer;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: indentation

File: src/ucp/tag/eager.c [hunk 7]
```diff
@@ -168,7 +169,7 @@ static size_t ucp_tag_pack_eager_last_contig(void *dest, void *arg)
     memcpy(hdr + 1, req->send.buffer + req->send.state.offset, length);
     return sizeof(*hdr) + length;
 }
-
+
```
Comment (yosefe) [inline src/ucp/tag/eager.c]: pls remove

File: test/gtest/ucp/test_ucp_memheap.cc [hunk 8]
```diff
@@ -7,6 +8,66 @@
...
+    ASSERT_UCS_OK(status);
+
+    std::string expected_data[300];
+    assert (max_iter <= 300);
+
+    for (int i = 0; i < max_iter; ++i) {
+        ucs_assert(size*i + alignment <= memheap_size);
+
+        (this->*send)(pe0, size, (void*)((uintptr_t)memheap + alignment + i * size),
+                        rkey, expected_data[i]);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_memheap.cc]: indentation

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: Good question. Since it is nonblocking shell we just return UCS_OK ?

File: src/ucp/core/ucp_request.h (hunk 2 above)
Comment (shamisp) [inline src/ucp/core/ucp_request.h]: I need to access this for both sends and rma, so I need it under one common struct.

File: src/ucp/rma/basic_rma.c (hunk 5 above)
Comment (shamisp) [inline src/ucp/rma/basic_rma.c]: missing or add ?

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (yosefe) [inline src/ucp/api/ucp.h]: I think we can return UCS_OK and UCS_INPROGRESS to be consistent with other functions, and possibly avoid calling ucp_worker_flush

File: src/ucp/core/ucp_request.h (hunk 2 above)
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: see my comment below..

File: src/ucp/rma/basic_rma.c (hunk 5 above)
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: missing == add :)
need to remove one

File: src/ucp/rma/basic_rma.c (hunk 4 above)
Comment (shamisp) [inline src/ucp/rma/basic_rma.c]: yes, we can. This is sort of styling preference.

File: src/ucp/core/ucp_ep.c [hunk 9]
```diff
@@ -61,14 +62,13 @@ void ucp_ep_destroy_uct_ep_safe(ucp_ep_h ep, uct_ep_h uct_ep)
...
 
-void ucp_ep_add_pending(ucp_ep_h ep, uct_ep_h uct_ep, ucp_request_t *req)
+void ucp_ep_add_pending(uct_ep_h uct_ep, ucp_request_t *req)
 {
     ucs_status_t status;
 
     ucs_trace_data("add pending request %p uct %p to uct_ep %p", req,
                    &req->send.uct, uct_ep);
 
-    req->send.ep = ep;
```
Comment (shamisp) [inline src/ucp/core/ucp_ep.c]: After restructuring I should revert this change

File: src/ucp/tag/eager.c [hunk 10]
```diff
@@ -1,5 +1,6 @@
 /**
  * Copyright (C) Mellanox Technologies Ltd. 2001-2015.  ALL RIGHTS RESERVED.
+ * Copyright (c) UT-Battelle, LLC. 2015. ALL RIGHTS RESERVED.
```
Comment (shamisp) [inline src/ucp/tag/eager.c]: Remove copyright

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1211/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1213/ for details.

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: agreed

**Comment by shamisp** [comment]
@yosefe - please review carefully. It was tested with RC and MM TLs

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1216/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1217/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1218/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/633/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/634/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/635/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1219/ for details.

File: src/ucp/core/ucp_request.h [hunk 11]
```diff
@@ -52,6 +53,17 @@ enum {
         } \
     }
 
+/*
+ * - Mark the request as completed
+ * - Otherwise - the request might be released - if so, return it to mpool.
+ */
+#define ucp_request_rma_complete(_req) \
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: why do we need another macro here? if needed we can add more debug info to the macro params of to the request itself

File: src/ucp/core/ucp_request.h [hunk 12]
```diff
@@ -89,11 +102,15 @@ typedef struct ucp_request {
...
+
+                struct {
+                    uint64_t      remote_addr; /* remote address */
+                    ucp_rkey_h    rkey;        /* Rkey */
+                } rma;
             };
 
             size_t                length;   /* Total length, in bytes */
             ucp_frag_state_t      state;
-            uct_pending_req_t     uct;
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: this was intentionally the last field, since it's less frequently used

File: src/ucp/rma/basic_rma.c [hunk 13]
```diff
@@ -75,6 +77,156 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
     return status;
 }
 
+static ucs_status_t ucp_progress_put_nbi(uct_pending_req_t *self)
+{
+
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: redundant spaceline

File: src/ucp/rma/basic_rma.c [hunk 14]
```diff
@@ -75,6 +77,156 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
...
+    req->flags = UCP_REQUEST_FLAG_RELEASED;
+    ucp_ep_add_pending(ep, ep->uct_ep, req);
+}
+
+ucs_status_t ucp_put_nbi(ucp_ep_h ep, const void *buffer, size_t length,
+                         uint64_t remote_addr, ucp_rkey_h rkey)
+{
+    ucs_status_t status;
+    uct_rkey_t uct_rkey;
+    size_t frag_length;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: frag_length is not needed

File: src/ucp/rma/basic_rma.c [hunk 15]
```diff
@@ -128,6 +280,98 @@ ucs_status_t ucp_get(ucp_ep_h ep, void *buffer, size_t length,
...
+{
+    ucs_status_t status;
+    size_t frag_length;
+
+    uct_rkey_t uct_rkey = UCP_RKEY_LOOKUP(ep, rkey);
+
+    UCP_RMA_CHECK_PARAMS(buffer, length);
+    uct_rkey = UCP_RKEY_LOOKUP(ep, rkey);
+
+    for (;;) {
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: there is code duplication of ucp_progress_get_nbi()

File: test/gtest/ucp/test_ucp_memheap.cc [hunk 16]
```diff
@@ -7,6 +8,134 @@
 #include "test_ucp_memheap.h"
 
 
+void test_ucp_memheap::test_nonblocking_implicit_stream_xfer(nonblocking_send_func_t send, size_t alignment)
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_memheap.cc]: please don't copy/paste the code from test_blocking xfer, but extract common code to a function

File: test/gtest/ucp/test_ucp_memheap.cc [hunk 17]
```diff
@@ -7,6 +8,134 @@
...
+
+    pe0->disconnect();
+    pe1->disconnect();
+
+    ucp_rkey_buffer_release(rkey_buffer);
+    status = ucp_mem_unmap(pe1->ucph(), memh);
+    ASSERT_UCS_OK(status);
+}
+
+void test_ucp_memheap::test_nonblocking_implicit_xfer(nonblocking_send_func_t send, size_t alignment)
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_memheap.cc]: again, please avoid copy/paste.

File: test/gtest/ucp/test_ucp_rma.cc [hunk 18]
```diff
@@ -22,6 +34,18 @@ class test_ucp_rma : public test_ucp_memheap {
...
 
+    void nonblocking_get_nbi(entity *e, size_t max_size, 
+                             std::string& local_data, void *memheap_addr,
+                             ucp_rkey_h rkey)
+    {
+        ucs_status_t status;
+
+        ucs::fill_random((char*)memheap_addr, (char*)memheap_addr + max_size);
+        status = ucp_get_nbi(e->ep(), (void *)&local_data[0], local_data.length(),
+                         (uintptr_t)memheap_addr, rkey);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_rma.cc]: indentation

File: src/ucp/rma/basic_rma.c [hunk 19]
```diff
@@ -75,6 +77,156 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
...
+            }
+
+            req->send.buffer += packed_len;
+            req->send.rma.remote_addr += packed_len;
+        } else {
+            break;
+        }
+    }
+
+    return status;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: it's not enough to return from pending function. You also have to make sure ucp_worker_flush() does not return before this happens (e.g pending requests are flushed). Please note that *UCT flush does not flush pending* to avoid possible deadlocks with 2-sided protocols. One way to do it is to add counter for outstanding RMAs.

File: test/gtest/ucp/test_ucp_memheap.cc [hunk 20]
```diff
@@ -7,6 +8,134 @@
...
+    pe1->connect(pe0);
+
+    ucp_rkey_h rkey;
+    status = ucp_ep_rkey_unpack(pe0->ep(), rkey_buffer, &rkey);
+    ASSERT_UCS_OK(status);
+
+    std::string expected_data[300];
+    assert (max_iter <= 300);
+
+    for (int i = 0; i < max_iter; ++i) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_memheap.cc]: please add a test which actually goes to pending (i guess max_iter should be bigger), the test should fail with the current code because flush will not actually flush it.

File: src/ucp/core/ucp_request.h (hunk 11 above)
Comment (shamisp) [inline src/ucp/core/ucp_request.h]: The other one call callback function that I don't need for RMA. If we merge the two macros it introduces branch in the critical path.

File: src/ucp/rma/basic_rma.c (hunk 15 above)
Comment (shamisp) [inline src/ucp/rma/basic_rma.c]: it is a bit similar but not code duplication. I tried to merge the two functions and it looked really ugly and over complex. This looks much cleaner.

File: src/ucp/rma/basic_rma.c (hunk 13 above)
Comment (shamisp) [inline src/ucp/rma/basic_rma.c]: ok

File: src/ucp/rma/basic_rma.c (hunk 14 above)
Comment (shamisp) [inline src/ucp/rma/basic_rma.c]: ok

File: test/gtest/ucp/test_ucp_memheap.cc (hunk 20 above)
Comment (shamisp) [inline test/gtest/ucp/test_ucp_memheap.cc]: the stream version does it

Comment (shamisp) [inline test/gtest/ucp/test_ucp_memheap.cc]: Surprisingly, it enters pending queue flow and invokes progress

File: test/gtest/ucp/test_ucp_rma.cc (hunk 18 above)
Comment (shamisp) [inline test/gtest/ucp/test_ucp_rma.cc]: ok

File: src/ucp/rma/basic_rma.c (hunk 19 above)
Comment (shamisp) [inline src/ucp/rma/basic_rma.c]: We have to discuss this. I don't want to maintain two many counters. We already have at least two on the uct level.

File: src/ucp/core/ucp_request.h (hunk 11 above)
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: i believe you can pass 'void' as the callback parameter to this macro to avoid calling a callback.

File: src/ucp/rma/basic_rma.c (hunk 19 above)
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: yes. lets discuss on dev call.

File: test/gtest/ucp/test_ucp_memheap.cc (hunk 20 above)
Comment (yosefe) [inline test/gtest/ucp/test_ucp_memheap.cc]: we have to discuss this. specifically, do we define that uct_flush also clears up the pending queue? currently AFAIK it's not defined, but looks like this what happens in practice.

Comment (shamisp) [inline test/gtest/ucp/test_ucp_memheap.cc]: Yep, let's target to talk tomorrow.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1248/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/661/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@sameerkm please take a look

File: src/ucp/core/ucp_request.h (hunk 11 above)
Comment (shamisp) [inline src/ucp/core/ucp_request.h]: I think it was crushing on me when I did it. I will double check

**Comment by shamisp** [comment]
@yosefe please take a look.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1254/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/667/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
This change was :+1:  by @sameerkm

File: src/ucp/rma/basic_rma.c (hunk 19 above)
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: colnclusion - uct flush ensures pending is also flushed

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1266/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/679/ for details (Mellanox internal link).
